kiwi/xml_state.py

```diff
@@ -1622,6 +1631,71 @@ def delete_repository_sections_used_for_build(self) -> None:
             ]
         )
 
+    def add_obs_repositories(
+        self, user: str, password: str, project: str, name: str,
+        profile: str, arch: str, repo: str
+    ) -> None:
+        """
+        Add repositories from the given obs project
+
+        :param str user: OBS account user name
+        :param str password: OBS account password
+        :param str project: OBS project path
+        :param str name: OBS image package name
+        :param str profile: OBS image multibuild profile name
+        :param str arch: OBS architecture
+        :param str repo: OBS image package build repository name
+        """
+        project_name = name if not profile else f'{name}:{profile}'
+        runtime_config = RuntimeConfig()
+        try:
+            buildinfo_link = os.sep.join(
+                [
+                    runtime_config.get_obs_api_server_url(),
+                    project, repo or 'images', arch or 'x86_64',
+                    project_name, '_buildinfo'
+                ]
+            )
+            request = requests.get(
+                buildinfo_link, auth=HTTPBasicAuth(user, password)
+            )
+            request.raise_for_status()
+        except Exception as issue:
+            raise KiwiUriOpenError(
+                f'{type(issue).__name__}: {issue}'
+            )
+
+        buildinfo_download = NamedTemporaryFile()
+        with open(buildinfo_download.name, 'wb') as buildinfo:
+            buildinfo.write(request.content)
+
+        buildinfo_xml_tree = etree.parse(buildinfo_download.name)
+        repo_paths = buildinfo_xml_tree.getroot().xpath(
+            '/buildinfo/path'
+        )
+        if not repo_paths:
+            raise KiwiOBSBuildInfoError(
+                f'OBS buildinfo for {project_name} has no repo paths'
+            )
+        for repo_path in repo_paths:
+            repo_url = repo_path.get('url')
+            if repo_url:
+                repo_check = SolverRepository.new(Uri(repo_url, 'rpm-md'))
+                use_repo = True
+                try:
+                    repo_check.timestamp()
+                except Exception as issue:
+                    # we will support real rpm-md repos only. The check for
+                    # the timestamp of the primary repodata object is an
+                    # indicator If the repo can be used, if not the repo
+                    # will be ignored.
+                    log.debug(
+                        f'Repo: {repo_url} is not a valid rpm-md type: {issue}'
+                    )
+                    use_repo = False
+                if use_repo:
+                    self.add_repository(repo_url, 'rpm-md')
```

Comment:
What about having some sort of increasing priority. So first has highest priority, this way the result is likely to be closer to what OBS is doing. This is specially sensitive when testing specific patches that relay on its own project repository being appended on top of the repos list.

Reply:
Oh yes this is a good idea. The problem is how should this be done consistently across the package managers. For zypper a low prio is 99 and a high prio is 1. So in this case it could be just "1 2 3 4". I think dnf is the same according to https://dnf.readthedocs.io/en/latest/conf_ref.html. How is it done for apt and pacman... still don't know

Reply:
Oh and it probably doesn't matter for non rpm-md repos because the current code only uses rpm-md repo, which could be a problem if the image is build for arch or debian based distros. Hmm a good question actually

Reply:
ok for an ubuntu build a path element could look like this

```xml
<path project="Ubuntu:20.04" repository="universe" url="https://download.opensuse.org/repositories/Ubuntu:/20.04/universe/"/>
```

And very interesting is that the provided URL is not reachable and gives a 404

Reply:
The Pin-Priority for apt works the other way round, low number == low prio

Reply:
ok so looking closer opens some questions for me:

* obs buildinfo reports urls which are not accessible from the outside like https://download.opensuse.org/repositories/Ubuntu:/20.04/universe/ but obs can build with them ??
* the prio setting is different per package manager, how to handle this
* the current code does a repo_check based on a timestamp. This only works for rpm-md repos because kiwi/solver/repository has no implementation for apt and pacman repotypes. Maybe there is a better check to find out the repo type ?

Reply:
Well I believe it doesn't really matter for apt, pacman and probably for dnf too, at least for SUSE's OBS instances. So I believe the key factor is that only published repositories are going to be functional if I understand it correctly. Meaning that third party repositories won't work and unpublished repositories too (in OBS they can be unpublished but still used for build).

Reply:
@davidcassany I think I can add a get_repotype method to the solver code. Can you help me with the identification ?
for rpm-md the presence of repodata is a good check, for apt I would look for Packages.gz and for pacman I would look for _Linux.db. Does this sound correct ? pacman would be worst because the meta data file names are not static. Thoughts ?

Reply:
>     * obs buildinfo reports urls which are not accessible from the outside like https://download.opensuse.org/repositories/Ubuntu:/20.04/universe/ but obs can build with them ??

I believe this is similar to what happens when using repositories that not published. OBS first tries to fetch them using the URL and in case it fails or it is not found then it tries to download them using the OBS API.

Reply:
> Well I believe it doesn't really matter for apt, pacman and probably for dnf too, at least for SUSE's OBS instances. So I believe the key factor is that only published repositories are going to be functional if I understand it correctly. Meaning that third party repositories won't work and unpublished repositories too (in OBS they can be unpublished but still used for build).

yes and the code will find out if the repo can be reached or not and will add a debug log. That should be better a warning message. I will change that. But we still need a better detection of the repo type. Only rpm-md is not enough because in obs you can decided if you publish or not as you said. For example someone building against Virt:App:Builder for ubuntu will use the apt repo with python3-kiwi in it which is also reachable publicly.

Reply:
I am thinking of another approach to do something similar. What if instead of making KIWI using some remote repositories we use some local ones created by `osc`. This way we use the same exact repositories and we are likely to achieve pretty equivalent results.

Gonna investigate if it is possible to make `osc` run the solver and build the repositories without actually build the image.

Reply:
or maybe easier, we let the package manager decide for the type and just add the repo if it's reachable. But obs also contains things like this

```xml
<path project="openSUSE:Tumbleweed" repository="dod" url="https://download.opensuse.org/repositories/openSUSE:/Tumbleweed/dod/"/>
```

which I also don't understand

Reply:
but this is also not public, so can be ignored. I'll update the code and then we need to find a way how to handle prios

Reply:
Yet another idea. The command `osc build --preload images x86_64` just runs the solver and creation of repositories. What if we code that as a new KIWI subcommand, so we could have an extra command that runs this preload. Probably it is not a crazy idea trying to code our own `osc build --preload images x86_64` to avoid `osc` dependencies. The solver is executed remotely as you did here and this PR. What if we also download the packages? The XML provides the list of packages (including arch, project and repo for each), so probably we can simply run OBS requests to download them.

Reply:
Hmm, so you are saying we create a new sub-command that does:

* send OBS api request to the get the _buildinfo
* use the _buildinfo data and download all packages. This would then be a _service:download_url: (not sure) api call
* collect all packages and create a repo out of them. Here I see a difficulty. for rpm-md it's a simple createrepo call but how to handle other repo-types ?
* let kiwi build with --set-repo _the-collective-repo_ instead of obsrepositories

Something like this ?

How to handle caching of downloaded packages ?

I'm not sure if we should do this. Actually some of the problems we don't like from the obs approach (that different solver and the re-creation of repos) would be again implemented but now in kiwi. I think this is not needed because if this is the plan people can use "osc build" with all its consequences and side-effects.

With this PR here my idea was to allow using the list of repositories from obs and all the rest is disconnected from obs as a service. I agree with you this requires the repos to be published and if they are not another option can be used. Options in this case are:

* find the public repos if they exist
* make use of "osc build"

But coding kiwi's own osc build implementation ties it strictly to the obs api, requires a credentials setup and re-implements quite some code from osc itself.

Do you think this is worth the effort ?

Reply:
```osc build --preload images x86_64``` for this to work you also have to checkout the project from obs, correct ?

Reply:
> `osc build --preload images x86_64` for this to work you also have to checkout the project from obs, correct ?

yes

Reply:
> How to handle caching of downloaded packages ?

IIRC we are already doing that. There is no caching for local repositories.

>     * collect all packages and create a repo out of them. Here I see a difficulty. for rpm-md it's a simple createrepo call but how to handle other repo-types ?

Yes that's a good point and an issue I underestimated, I was thinking of plain rpm-dir, but yes this is likely to be not sufficient in many cases.

> I think this is not needed because if this is the plan people can use "osc build" with all its consequences and side-effects.

We are approaching this feature slightly different and I see your point. My idea was more in the direction of `osc build` is not trustable and sometimes fails locally and it doesn't remotely, so what if we find and functional alternatively.

> But coding kiwi's own osc build implementation ties it strictly to the obs api, requires a credentials setup and re-implements quite some code from osc itself.

Yes that's true. Yes the real issue of this is the repo creation which is package manager specific, I was just thinking of only downloading binaries from OBS using some simple structured folder setup <repo>/<package>, but yes this is not sufficient and the missing part is getting into the burden of OSB build procedures.

Reply:
@davidcassany ok I've added a very simple get_repo_type() method that tries to lookup repo metada on the server. The implementation is poor for Arch as they don't seem to have static repo metadata files like anyone else has.

Together with the information about the repo type I added the repo priority setting as you suggested.

Given the repos are available publicly (or can be accessed by the configured OBS download_url in the kiwi.yml) this feature can be used to build with the obs repos and prios

BUT

I'm not sure if we should merge it